hedyorg/hedy
# 🖊️ Allow if_pressed to work with functions

This PR allows the if-pressed feature to be used with functions. Specifically, it introduces the following set of changes:
- If a program does not contain an if-pressed statement, its transpilation remains unchanged
- If a program contains an if-pressed statement it will use scopes:
    - a global scope dictionary is added to the start of the program
    - a local scope dictionary is added to the start of every function. If the function has any arguments, the are added to the local scope.
    - when a variable is assigned (appears on the lhs of an assignment), instead of creating a new variable, we add a key to the current scope. If the assignment is in the global scope, we use the global scope. If it is in a function, we use the specific local scope.
    - when for-list and for-loops are used, the iterator variable (`for a in animals`) is not added no any scope (since it must be accessible only in the scope of the loop).
    - when a variable is retrieved (e.g. appears on the rhs of an assignment or in a print statement), we try to fetch it from the local scope, then from the global scope, and then directly as a variable. The last case is required to support the variables defined in for-loops (`for a in animals`).

Fixes #5749

**How to test**
All scenarios I could think of are covered with automated tests. However, they are far from exhaustive. Please use the following scenarios as starting points to explore different combinations, i.e. just try to break it:
- if-pressed correctly should reassign a variable from outer scope:
```
# level 12
x is 'a'
if x is pressed
  x is 'great'
else
  x is 'not great'
print x
```
- if-pressed should expose a variable created within it to the outer scope:
```
# level 12
x is 'a'
if x is pressed
  m is 'great'
else
  m is 'not great'
print m
```
- if-pressed should work within a function:
```
# level 12
define make_turn
    if x is pressed
        print 'good'
    else
        print 'bad'
call make_turn
```
- if-pressed should work within a function with an argument:
```
# level 12
define make_turn with x
    if x is pressed
        print 'good'
    else
        print 'bad'
call make_turn with 'm'
```
- if-pressed should work outside of a function:
```
# level 17
define go_left with last_direction:
    if last_direction = 'right':
        turn 180

define go_right with last_direction:
    if last_direction = 'left':
        turn 180

move = True
last_direction = 'left'

while move = True:
    if a is pressed:
        call go_left with last_direction
        last_direction = 'left'
    elif d is pressed:
        call go_right with last_direction
        last_direction = 'right'
    else:
        move = False
    if move is True:
        forward 25
```
- If-pressed should work within for-loops:
```
# level 12
animals is 'cat', 'dog', 'platypus'
if a is pressed
    for animal in animals
        print animal
else
    print 'no animals'
```
